### DISCUSSION TITLE
Failed to read SMTP TLS certificate

### DISCUSSION THREAD

**Discussion by mdobydullah** [postalserver/postal#1512] [Help with using Postal] [answered]
Hi, I set up Postal with docker (followed: https://docs.postalserver.io/) and added cert files to the postal.yml file.

```
smtp_server:
  tls_enabled: true
  tls_certificate_path: /opt/postal/config/smtp.crt
  tls_private_key_path: /opt/postal/config/smtp.key
```

Tested with: https://www.gmass.co/smtp-test

Results:
```
<< 250-My capabilities are
<< 250-STARTTLS
<< 250 AUTH CRAM-MD5 PLAIN LOGIN
>> STARTTLS
<< 220 Ready to start TLS
ERROR: An error occurred while attempting to establish an SSL or TLS connection.
```

Then I checked Postal SMTP log:

![ssl file not found](https://user-images.githubusercontent.com/13184472/127612232-bd5aff50-de9f-47ee-9d1d-dd573b0fcecf.png)

How can I solve this?

**Comment by willpower232** [comment]
I think you should use `postal bash smtp` to connect to the smtp container and then verify that the certificate files are present and readable

**Reply by mdobydullah** [reply]
Thanks @willpower232 for quick response.

The certs are present in the location. I'm going to test with ```postal bash smtp```.

![postal ssl key](https://user-images.githubusercontent.com/13184472/127627052-3a09c90e-4298-4898-ba6e-f06ae920c449.png)

After solving this issue, I can make my email server live.

**Reply by willpower232** [reply]
Its always worth using `postal bash` to verify that any files have made their way into the docker container. You can also verify the permissions on the file that way.

**Reply by mdobydullah** [reply]
I cannot access certs from ```postal bash smtp```. After exiting ```postal bash smtp```, I can access the cert files.

Where can I store my certs? Could you please tell me the right location to store cert files?

Also, I've installed Caddy and SSL is working. Is there any way to use Caddy SSL certs for SMTP ?

**Reply by willpower232** [reply]
It seems like you have the files in the right place so you could try `postal stop` and `postal start` to see if that sorts docker out.

**Reply by mdobydullah** [reply]
Don't work. I always run this command ```postal stop && postal start``` after updating the config file.

**Reply by mdobydullah** [reply]
At last works. I found a temp solution. 

I run ```postal bash smtp``` and created a custom folder named **certs** and stored cert files in the **certs** directory. After that it works.

**postal.yml:**

```
smtp_server:
  port: 25
  tls_enabled: true
  tls_certificate_path: /opt/postal/app/certs/smtp.crt
  tls_private_key_path: /opt/postal/app/certs/smtp.key
```

Now TLS is working...! Still, there is an issue.

**The Issue:**
After restarting postal, the custom folders and files are removed from ```postal bash smtp``` location. I need to create **certs** folder and cert files again. You should work on this issue.

Anyway, thank you so much @willpower232. :heartpulse:

**Reply by willpower232** [reply]
You can see in docker-compose.yml that the /opt/postal/config folder is linked into each docker container https://github.com/postalserver/install/blob/main/docker-compose.yml

But actually its linked to /config which means the  config file should be 
```
  tls_certificate_path: /config/smtp.crt
  tls_private_key_path: /config/smtp.key
  ```

Can you confirm @adamcooke ?

**Reply by mdobydullah** [reply]
Yeah. It works. 💗

I tried with full paths and didn't work:

```
tls_certificate_path: /opt/postal/config/smtp.crt
tls_private_key_path: /opt/postal/config/smtp.key
```

Now it works with this config and paths: 

```
smtp_server:
  port: 25
  tls_enabled: true
  tls_certificate_path: /config/smtp.crt
  tls_private_key_path: /config/smtp.key
```

@willpower232 Thank you so much, again. :blush:

**Reply by willpower232** [reply]
actually if you're using the default filenames, I think you can probably remove those two lines of config and it would work by default

**Reply by mdobydullah** [reply]
Yeah. Thanks again, bro. 😊

**Reply by willpower232** [reply]
Nice, I'll remember this for when my time to upgrade comes!

**Reply by mdobydullah** [reply]
Great. You can also write a short note in docs (https://docs.postalserver.io/) about SMTP TLS configuration. It will be very helpful. It takes almost 2 days to fix this small issue.

**Reply by willpower232** [reply]
Absolutely, I haven't had a chance to read through it all yet but its on my list :sweat_smile:

**Reply by mdobydullah** [reply]
I already read the whole docs a few times. 😅

Anyway, thanks man!

**Reply by XenyFord** [reply]
> 
> 
> Its always worth using `postal bash` to verify that any files have made their way into the docker container. You can also verify the permissions on the file that way.

Hi Mr. Will,

I tried both the ways with `/config/smtp.crt` and `/opt/postal/config/smtp.crt` but, still, I'm getting the same error "smtp_1 | [10] [2021-09-01T18:57:48.551] ERROR -- : [UDBNBP] Errno::ENOENT: No such file or directory @ rb_sysopen - /config/smtp.crt"

![131729905-d87251d7-54e1-4ae1-b98d-0611ccc682a9](https://user-images.githubusercontent.com/89290310/132002249-932a3b52-faef-41da-8b9a-8a8e5070497f.jpg)

Could you please help me solve this? Thank you in advance!

**Reply by willpower232** [reply]
@XenyFord can you stop commenting with the same question in random places? its very hard to see where you have actually received help

**Reply by XenyFord** [reply]
Mr. Will,
Okay. Absolutely, I will not do that. Sorry for bothering you!

**Reply by willpower232** [reply]
I can't find where you first asked about this but I think you want to try https://github.com/postalserver/postal/discussions/1570#discussioncomment-1270528

**Comment by adamcooke** [comment, marked as answer]
When you're reference files in `/opt/postal/config` in the `postal.yml`, you need to remember that the containers that are running Postal don't have access to operating system's file system. The `/opt/postal/config` directory is mounted at `/config`. 

Assuming the files are in `/opt/postal/config` on the host, you can just do this.

```yaml
smtp_server:
  port: 25
  tls_enabled: true
  tls_certificate_path: /config/smtp.crt
  tls_private_key_path: /config/smtp.key
```

**Reply by mdobydullah** [reply]
Thank you so much, @adamcooke.

@willpower232 helped me to fix the issue.

**Reply by XenyFord** [reply]
> 
> 
> Thank you so much, @adamcooke.
> 
> @willpower232 helped me to fix the issue.

Hello Mr. Obydullah,

Did you find the solution for sending emails through the secured connection via TLS/SSL ?
I'm still getting the same error "smtp_1      | [10] [2021-09-01T18:57:48.551] ERROR -- : [UDBNBP] Errno::ENOENT: No such file or directory @ rb_sysopen - /config/smtp.crt"

Could you please help me solve this? Thank you in advance!

![Screenshot_1](https://user-images.githubusercontent.com/89290310/131729905-d87251d7-54e1-4ae1-b98d-0611ccc682a9.jpg)

**Reply by hswamy** [reply]
@XenyFord Please check the permission for smtp.crt, ( Give read access )

**Reply by XenyFord** [reply]
Hi Sir,
I'm not able to find such file named "smtp.crt". Could you please let me know how to find and apply the same in right way? 
Thank you in advance for your help.
